src/libexpr/primops.cc

```diff
@@ -513,6 +515,97 @@ static RegisterPrimOp primop_isFunction({
     .fun = prim_isFunction,
 });
 
+void prim_isThunk(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    auto &arg = *args[0];
+    v.mkBool(arg.isThunk());
+}
+
+static RegisterPrimOp primop_isThunk({
+    .name = "isThunk",
+    .args = {"v"},
+    .doc = R"(
+      Returns `true` if the argument is a thunk, `false` otherwise.
+
+      This is internal, because the thunk-ness of a value is not a stable property, while the Nix language is meant to be referentially transparent.
+    )",
+    .fun = prim_isThunk,
+    .internal = true
+});
+
+void prim_printValue(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    auto &arg = *args[0];
+    std::stringstream ss;
+    printValue(state, ss, arg);
+    v.mkString(ss.str());
+}
+
+static RegisterPrimOp primop_printValue({
+    .name = "printValue",
+    .args = {"v"},
+    .doc = R"(
+      Print any value to an informative string.
+
+      This is internal, because the format is not a stable behavior, while the Nix language is meant to be stable over time.
+    )",
+    .fun = prim_printValue,
+    .internal = true
+});
+
+
+/* Determine whether the argument is an integer. */
+static void prim_seqWithCallPos(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    auto p = args[0];
+    auto f = args[1];
+    auto a = args[2];
+    state.forceValue(*args[0], pos);
+    state.forceValue(*args[1], pos);
+    auto intro_args = state.buildBindings(4);
+    auto pos_attrs = state.allocValue();
+    intro_args.insert(state.symbols.create("pos"), pos_attrs, noPos);
+
+    auto vIsThunk = get(state.internalPrimOps, "isThunk");
+    assert(vIsThunk);
+    intro_args.insert(state.symbols.create("isThunk"), *vIsThunk, noPos);
+
+    auto vPrintValue = get(state.internalPrimOps, "printValue");
+    assert(vPrintValue);
+    intro_args.insert(state.symbols.create("printValue"), *vPrintValue, noPos);
+
+    auto [rows,cols] = getWindowSize();
+    auto vCols = state.allocValue();
+    vCols->mkInt(cols);
+    intro_args.insert(state.symbols.create("terminalWidth"), vCols , noPos);
```

Comment:
We discussed this builtin in the PR and propose to leave out the terminalWidth for the first iteration.